subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -262,64 +242,73 @@ DefaultJavaCompileSpec createSpec() {
             compileOptions.setSourcepath(getProjectLayout().files(sourcesRoots));
         }
         spec.setAnnotationProcessorPath(compileOptions.getAnnotationProcessorPath() == null ? ImmutableList.of() : ImmutableList.copyOf(compileOptions.getAnnotationProcessorPath()));
-        configureCompatibilityOptions(spec);
+        configureCompileOptions(spec);
         spec.setSourcesRoots(sourcesRoots);
 
-        if (!toolchainCompatibleWithJava8) {
+        if (!isToolchainCompatibleWithJava8()) {
             spec.getCompileOptions().setHeaderOutputDirectory(null);
         }
         return spec;
     }
 
+    private void validateForkOptionsMatchToolchain() {
```

Comment:
Likewise with the validations for the toolchains - co-locating these in a common static util class would be helpful.

Reply:
Same here, will address in later PRs.